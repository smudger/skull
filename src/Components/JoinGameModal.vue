<script setup>
import { DialogTitle } from "@headlessui/vue";
import BaseModal from "@/Components/Base/BaseModal.vue";

defineProps({
  show: { type: Boolean, required: true },
});

const emit = defineEmits(["join"]);

const joinGame = (event) => {
  const name = event.target.elements.name.value;

  emit("join", name);
};
</script>

<template>
  <BaseModal :show="show">
    <div class="text-center">
      <form @submit.prevent="joinGame">
        <DialogTitle
          as="h3"
          class="text-base font-semibold leading-6 text-gray-900"
          >What shall we call you?</DialogTitle
        >
        <div class="mt-2">
          <label class="sr-only" for="name">Name</label>
          <div class="mt-2">
            <input
              id="name"
              class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-rose-600 sm:text-sm sm:leading-6"
              name="name"
              required
            />
          </div>
        </div>
        <div class="mt-5 sm:mt-6">
          <button
            class="inline-flex w-full justify-center rounded-md bg-rose-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-rose-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-rose-600"
            type="submit"
          >
            Join
          </button>
        </div>
      </form>
    </div>
  </BaseModal>
</template>
