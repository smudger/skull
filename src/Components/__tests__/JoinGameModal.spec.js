import { afterAll, beforeAll, describe, expect, it } from "vitest";
import { config, mount } from "@vue/test-utils";
import JoinGameModal from "@/Components/JoinGameModal.vue";
import BaseModal from "@/Components/Base/BaseModal.vue";

const decks = [
  {
    name: "Pink",
    bgColor: "bg-pink-500",
    selectedColor: "ring-pink-500",
    isAssigned: false,
  },
  {
    name: "Purple",
    bgColor: "bg-purple-500",
    selectedColor: "ring-purple-500",
    isAssigned: false,
  },
  {
    name: "Blue",
    bgColor: "bg-blue-500",
    selectedColor: "ring-blue-500",
    isAssigned: false,
  },
  {
    name: "Green",
    bgColor: "bg-green-500",
    selectedColor: "ring-green-500",
    isAssigned: false,
  },
  {
    name: "Yellow",
    bgColor: "bg-yellow-500",
    selectedColor: "ring-yellow-500",
    isAssigned: true,
  },
];

describe("JoinGameModal", () => {
  beforeAll(() => {
    config.global.renderStubDefaultSlot = true;
  });

  afterAll(() => {
    config.global.renderStubDefaultSlot = false;
  });

  it("is visible when the show prop is true", () => {
    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks },
    });

    expect(wrapper.findComponent(BaseModal).props().show).toBe(true);
  });

  it("is not visible when the show prop is false", () => {
    const wrapper = mount(JoinGameModal, {
      props: { show: false, decks },
    });

    expect(wrapper.findComponent(BaseModal).props().show).toBe(false);
  });

  it("emits a join event when the form is submitted", async () => {
    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks },
      global: {
        stubs: {
          BaseModal: true,
          DialogTitle: true,
        },
      },
    });

    await wrapper.find("input").setValue("Constance Wu");

    await wrapper.find("form").trigger("submit.prevent");

    expect(wrapper.emitted()).toHaveProperty("join");

    expect(wrapper.emitted("join")[0][0]).toStrictEqual({
      name: "Constance Wu",
      deck: decks[0].name,
    });
  });

  it("requires you to choose a name", async () => {
    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks },
      global: {
        stubs: {
          BaseModal: true,
          DialogTitle: true,
        },
      },
    });

    expect(wrapper.find("input").attributes()).toHaveProperty("required");
  });

  it("renders an option for each deck color", async () => {
    const blueAndGreen = [
      { name: "Blue", bgColor: "bg-blue-500", selectedColor: "ring-blue-500" },
      {
        name: "Green",
        bgColor: "bg-green-500",
        selectedColor: "ring-green-500",
      },
    ];

    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks: blueAndGreen },
      global: {
        stubs: {
          BaseModal: true,
          DialogTitle: true,
        },
      },
    });

    const options = wrapper.findAll('div[role="radio"]');

    expect(options).toHaveLength(2);
    expect(options[0].text()).toBe("Blue");
    expect(options[1].text()).toBe("Green");
  });

  it("selects the first unassigned deck by default", async () => {
    const blueAndGreen = [
      {
        name: "Blue",
        bgColor: "bg-blue-500",
        selectedColor: "ring-blue-500",
        isAssigned: true,
      },
      {
        name: "Green",
        bgColor: "bg-green-500",
        selectedColor: "ring-green-500",
        isAssigned: false,
      },
    ];

    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks: blueAndGreen },
      global: {
        stubs: {
          BaseModal: true,
          DialogTitle: true,
        },
      },
    });

    const options = wrapper.findAll('div[role="radio"]');
    expect(options[0].attributes()).toHaveProperty("aria-checked", "false");
    expect(options[1].attributes()).toHaveProperty("aria-checked", "true");
  });

  it("disables the assigned decks for selection", async () => {
    const blueAndGreen = [
      {
        name: "Blue",
        bgColor: "bg-blue-500",
        selectedColor: "ring-blue-500",
        isAssigned: true,
      },
      {
        name: "Green",
        bgColor: "bg-green-500",
        selectedColor: "ring-green-500",
        isAssigned: false,
      },
    ];

    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks: blueAndGreen },
      global: {
        stubs: {
          BaseModal: true,
          DialogTitle: true,
        },
      },
    });

    const options = wrapper.findAll('div[role="radio"]');
    expect(options[0].attributes()).toHaveProperty("aria-disabled", "true");
    expect(options[1].attributes()).not.toHaveProperty("aria-disabled");
  });

  it("displays a warning when all the decks are assigned", async () => {
    const blueAndGreen = [
      {
        name: "Blue",
        bgColor: "bg-blue-500",
        selectedColor: "ring-blue-500",
        isAssigned: true,
      },
      {
        name: "Green",
        bgColor: "bg-green-500",
        selectedColor: "ring-green-500",
        isAssigned: true,
      },
    ];

    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks: blueAndGreen },
      global: {
        stubs: {
          BaseModal: true,
          DialogTitle: true,
        },
      },
    });

    expect(wrapper.find("form").exists()).toBe(false);
    expect(wrapper.text()).toContain("This game is full!");
    expect(wrapper.text()).toContain(
      "All the decks have been taken, but you can still join as a spectator.",
    );
  });

  it("renders correctly when visible", () => {
    const wrapper = mount(JoinGameModal, {
      props: { show: true, decks },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
