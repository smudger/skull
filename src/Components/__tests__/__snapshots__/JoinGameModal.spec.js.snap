// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JoinGameModal > renders correctly when visible 1`] = `
<base-modal-stub
  show="true"
>
  <div
    class="text-center"
  >
    <form
      name="<input id=\\"name\\" class=\\"block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-rose-600 sm:text-sm sm:leading-6\\" name=\\"name\\" required=\\"\\">"
    >
      <dialog-title-stub
        as="h3"
        class="text-base font-semibold leading-6 text-gray-900"
        id="headlessui-dialog-title-5"
      >
        What shall we call you?
      </dialog-title-stub>
      <div
        class="mt-2"
      >
        <label
          class="sr-only"
          for="name"
        >
          Name
        </label>
        <div
          class="mt-2"
        >
          <input
            class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-rose-600 sm:text-sm sm:leading-6"
            id="name"
            name="name"
            required=""
          />
        </div>
      </div>
      <div
        class="mt-5 sm:mt-6"
      >
        <button
          class="inline-flex w-full justify-center rounded-md bg-rose-600 px-3 py-2 text-sm font-semibold text-white shadow-sm hover:bg-rose-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-rose-600"
          type="submit"
        >
           Join 
        </button>
      </div>
    </form>
  </div>
</base-modal-stub>
`;
